design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+# Secure Communication between Contour and Envoy
+
+_Status_: Draft, in review.
+
+This document outlines how to secure the communication between Contour
+and Envoy containers.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy
+- Authenticate Envoy connecting to Contour
+- Rotating and invalidating the credentials
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
```

Comment:
Is this a separate init container (so a 2nd one)? It might be best to just have `bootstrap` do both tasks, creating the secrets and then the bootstrap config. I'm not sure why we'd want to create certs outside of the bootstrap unless this was a CLI command to set up a cluster (i.e. run from an admin workstation and not inside an init container).

Reply:
I believe there are use cases where the users might not run `contour bootstrap` and provide the bootstrap configuration separately. I thought it would be a better idea to have it a separate init container for better control on these cases.